### ISSUE TITLE
Create GCC 8.3 PR Testing Build

### ISSUE THREAD

**Issue by ZUUL42** [trilinos/Trilinos#6295]
## Enhancement
@trilinos/framework

Look into creating a PR Testing build that uses GCC 8.3.
Currently we have a GCC 7.2 PR testing build.
GCC 8.3 would likely replace the 7.2 build.

**Comment by ZUUL42** [comment]
A ticket has been submitted with SEMS to add a GCC 8.3 module.

Testing will look at how building with the same Werror flags the 7.2.0 build uses currently.

**Comment by bartlettroscoe** [comment]
Who is the customer for this?

**Comment by mhoemmen** [comment]
@bartlettroscoe Sierra would like a GCC 8.3 build, I think.  @rrdrake can comment more, but I think the reason is that GCC 8.3 has different (probably more) warnings.

**Comment by rrdrake** [comment]
The warnings from a newer GCC version is one reason, but also important is Astra (the Vanguard series of ARM machines) uses GCC 8 headers by default.

**Comment by ZUUL42** [comment]
[CDash report](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=5952912&filtercount=1&showfilters=1&field1=buildstamp&compare1=61&value1=20191123-0034-Experimental&filtercombine=and) from a run with GCC 8.3 @ Werror flags set as they have been in 7.2.0.
There are a number of Werror flags that are causing build errors.
@trilinos/muelu @trilinos/intrepid @trilinos/stk @trilinos/stokhos @trilinos/amesos2 @trilinos/zoltan2 @trilinos/ml @trilinos/panzer @trilinos/tpetra 
[gcc_8.3 branch](https://github.com/trilinos/Trilinos/compare/develop...ZUUL42:gcc_8.3)

**Comment by etphipp** [comment]
In the stokhos case all of the warnings/errors are coming from boost:
```
/projects/sems/install/rhel7-x86_64/semsboost/1.63.0/gcc/8.3.0/base/include/boost/mpl/assert.hpp:188:21: error: unnecessary parentheses in declaration of ‘assert_arg’ [-Werror=parentheses]
 failed ************ (Pred::************
                     ^
/projects/sems/install/rhel7-x86_64/semsboost/1.63.0/gcc/8.3.0/base/include/boost/mpl/assert.hpp:193:21: error: unnecessary parentheses in declaration of ‘assert_not_arg’ [-Werror=parentheses]
 failed ************ (boost::mpl::not_<Pred>::************
                     ^
cc1plus: all warnings being treated as errors
```
I don't think there is anything I can do about that.

**Comment by rppawlo** [comment]
The panzer build error is a "warning as error" coming from the boost library. Panzer does not actually use boost, but this header is being brought in through a panzer->percept->stk->boost dependency. We will have to wait for stk to clean this one out. (@alanw0 - no pressure, just cc'ing you). 

```
In file included from /projects/sems/install/rhel7-x86_64/semsboost/1.63.0/gcc/8.3.0/base/include/boost/geometry/core/closure.hpp:22,
                 from /projects/sems/install/rhel7-x86_64/semsboost/1.63.0/gcc/8.3.0/base/include/boost/geometry/geometry.hpp:25,
                 from /projects/sems/install/rhel7-x86_64/semsboost/1.63.0/gcc/8.3.0/base/include/boost/geometry.hpp:17,
                 from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_8.3.0/Trilinos/packagesstk_search/stk_search/BoostRTreeInterface.hpp:14,
                 from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_8.3.0/Trilinos/packagesstk_search/stk_search/CoarseSearchBoostRTree.hpp:40,
                 from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_8.3.0/Trilinos/packagesstk_search/stk_search/CoarseSearch.hpp:43,
                 from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_8.3.0/Trilinos/packages/perceptpercept/function/FieldFunction.hpp:35,
                 from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_8.3.0/Trilinos/packages/perceptadapt/UniformRefinerPattern.hpp:38,
                 from /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_8.3.0/Trilinos/packages/panzer/adapters-stkstk_interface/Panzer_STK_Interface.cpp:72:
/projects/sems/install/rhel7-x86_64/semsboost/1.63.0/gcc/8.3.0/base/include/boost/mpl/assert.hpp:188:21: error: unnecessary parentheses in declaration of ‘assert_arg’ [-Werror=parentheses]
 failed ************ (Pred::************
                     ^
/projects/sems/install/rhel7-x86_64/semsboost/1.63.0/gcc/8.3.0/base/include/boost/mpl/assert.hpp:193:21: error: unnecessary parentheses in declaration of ‘assert_not_arg’ [-Werror=parentheses]
 failed ************ (boost::mpl::not_<Pred>::************
                     ^
cc1plus: all warnings being treated as errors
```

**Comment by mhoemmen** [comment]
There's 1 Tpetra warning:
```
error: ‘void* memcpy(void*, const void*, size_t)’ writing to an object of non-trivially copyable type ‘using value_type = class Kokkos::complex<double>’ {aka ‘class Kokkos::complex<double>’}; use copy-assignment or copy-initialization instead [-Werror=class-memaccess]
       memcpy (outBuf, inBuf, numBytes);
       ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
```
The compiler is correct here -- `Kokkos::complex` is not [TriviallyCopyable](https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable).  I think this will be fixed in Kokkos -- it may already have been fixed.

**Comment by ZUUL42** [comment]
@rrdrake, what are the Werror flags that Sierra is using for their GCC 8 build?
Did you have similar issues with Boost WErrors.

**Comment by rrdrake** [comment]
The Werror flags for GCC 8 vary a little by library that we build, but a sample is
`-std=c++11 -fno-strict-aliasing -Wall -Wno-clobbered -Wno-vla -Wno-pragmas -Wno-unknown-pragmas -Wno-parentheses -Wno-unused-local-typedefs -Wno-literal-suffix -Wno-deprecated-declarations -Wno-misleading-indentation -Wno-int-in-bool-context -Wno-maybe-uninitialized -Wno-cast-function-type -Wno-class-memaccess -Wno-inline -Werror`

Yes, we are seeing some warnings coming out of boost.  Note that the Sierra TK team is slowly working to remove boost usage from stk mesh.

**Comment by etphipp** [comment]
Note that `-Wno-parentheses` turns off the warning that we are seeing from boost in Stokhos and Panzer.

**Comment by kddevin** [comment]
@ZUUL42 we haven't forgotten this issue; there is a discussion going in https://github.com/kokkos/kokkos/issues/2577 .  What is your timeframe for completing the new build?

**Comment by ZUUL42** [comment]
@kddevin, I don't have a set time frame.
But when it _is_ ready, I can get it right into the PR test set. I've got a possible build branch ready to go.
So, I'm just looking into how to fix issues preventing a successful GCC 8.3 Werror build and retesting as I see other changes make it into develop.
But I've got plenty else to work on as well.

**Comment by lucbv** [comment]
@ZUUL42 why not create a nightly test so that we can see progress on this work instead of running  it a PR only once in a while which makes it a bit hard to now if things are moving forward?

**Comment by alanw0** [comment]
Sorry for the delayed response... yes the stk team is working to remove boost dependence. Unfortunately it's a slow journey... But it is important to multiple stakeholders and we are making steady progress.

**Comment by ZUUL42** [comment]
@lucbv I am looking into some automated way to get you results on this work.
Until we get that setup I'll fire off a quick build every evening and post a link to the CDash report.

**Comment by ZUUL42** [comment]
I ran a test build just to make sure everything builds cleanly with Werror off. That looks good.
So the remainder is entirely 8.3 WError and that is 2 types, shadow and catch-value.
[Here](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=6015146&filtercount=1&showfilters=1&field1=buildstamp&compare1=63&value1=Experimental&filtercombine=and) is the last build as of 2019 Dec 6 11:52.

This is working with this set of WError flags:
`set(CMAKE_CXX_FLAGS "-std=c++11 -fno-strict-aliasing -Wall -Wno-clobbered -Wno-vla -Wno-pragmas -Wno-unknown-pragmas -Wno-parentheses -Wno-unused-local-typedefs -Wno-literal-suffix -Wno-deprecated-declarations -Wno-misleading-indentation -Wno-int-in-bool-context -Wno-maybe-uninitialized -Wno-cast-function-type -Wno-class-memaccess -Wno-inline -Wno-nonnull-compare -Wno-address -DTRILINOS_HIDE_DEPRECATED_HEADER_WARNINGS -Werror" CACHE STRING "Warnings as errors settings")`

**Comment by ZUUL42** [comment]
Just kicked off a new 8.3 test build.
[CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=61&value1=ascic166&field2=buildname&compare2=61&value2=PR-gcc_8.3-test-Trilinos_pullrequest_gcc_8.3.0-15&field3=buildstamp&compare3=61&value3=20191211-0033-Experimental)

**Comment by lucbv** [comment]
@ZUUL42 any chance you could fire another build?

**Comment by ZUUL42** [comment]
[New build](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=61&value1=ascic166&field2=buildname&compare2=61&value2=PR-gcc_8.3-test-Trilinos_pullrequest_gcc_8.3.0-16&field3=buildstamp&compare3=61&value3=20191216-1609-Experimental)

**Comment by ZUUL42** [comment]
I've broken it down to set WError on a package by package basis when building with GCC 8.3.0.
Everything builds without error _without_ WError set. Then _with_ WError set with the following flags:
```
-std=c++11
-fno-strict-aliasing 
-Wall 
-Wno-clobbered 
-Wno-vla 
-Wno-pragmas 
-Wno-unknown-pragmas 
-Wno-parentheses 
-Wno-unused-local-typedefs 
-Wno-literal-suffix 
-Wno-deprecated-declarations 
-Wno-misleading-indentation 
-Wno-int-in-bool-context 
-Wno-maybe-uninitialized 
-Wno-cast-function-type 
-Wno-class-memaccess 
-Wno-inline 
-Wno-nonnull-compare 
-Wno-address 
-DTRILINOS_HIDE_DEPRECATED_HEADER_WARNINGS
```
... we get these results:

The [first CDash report](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=6097274&filtercount=1&showfilters=1&field1=buildstamp&compare1=63&value1=Experimental&filtercombine=and) shows the GCC 8.3.0 WError issues with:
@trilinos/amesos2 
@trilinos/ifpack 
@trilinos/ml 
@trilinos/panzer 
@trilinos/rol 
@trilinos/rythmos 
@trilinos/sacado 
@trilinos/seacas 
@trilinos/thyra 
@trilinos/tpetra 

[This CDash report](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=6097220&filtercount=1&showfilters=1&field1=buildstamp&compare1=63&value1=Experimental&filtercombine=and) is for the @trilinos/intrepid GCC 8.3.0 WError issues.

**Comment by bartlettroscoe** [comment]
Why not just merge the files:

```
cmake/std/PullRequestLinuxGCC8.3.0TestingSettings.cmake
cmake/std/sems/PullRequestGCC8.3.0TestingEnv.sh
```

from the branch:

* https://github.com/trilinos/Trilinos/compare/develop...ZUUL42:gcc_8.3

to the 'develop' branch ASAP?  There is zero harm in having them in the 'develop' branch and that would make it trivial for developers to reproduce this build on the tip of 'develop'.

They you should set up a nightly build for this configuration with -Werror set.  That way, it works against the tip of 'develop' and people can watch every day to see the errors go away?

Really you should add this as a PR build and set -Werror on packages that are already clean to make sure they don't backslide as well.

**Comment by etphipp** [comment]
When I click on the CDash link I get this error:
```
HTTP/1.0 200 OK Cache-Control: no-cache, private Date: Mon, 06 Jan 2020 15:10:07 GMT You cannot access this project
```
Does anyone else see this?

**Comment by rppawlo** [comment]
@etphipp - I'm seeing the same thing.

**Comment by jmgate** [comment]
I see the same thing.

**Comment by ZUUL42** [comment]
Ya, it looks like the old results aren't accessible anymore.
I just got back in today. I'll get some new results posted for everything.

**Comment by ZUUL42** [comment]
Here are some fresh [test build results](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=gcc_8.3.0&field2=buildstamp&compare2=63&value2=Experimental).

**Comment by ZUUL42** [comment]
New test build started.
[CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=gcc_8.3.0&field2=buildstamp&compare2=63&value2=Experimental)

**Comment by lucbv** [comment]
@ZUUL42 I have a flurry of PR ongoing that should resolve a bunch of warnings in the gcc/8.3.0 build so if you wouldn't mind launching a build of your own tonight so that we can see the results tomorrow, that would be nice.

**Comment by ZUUL42** [comment]
@lucbv, that all looks pretty good to me. Thanks.
I kicked off a build, but I would guess we're down to just the two last Werrors in @trilinos/sacado & @trilinos/seacas. [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=61&value1=ascic166&field2=buildname&compare2=61&value2=PR-gcc_8.3-test-Trilinos_pullrequest_gcc_8.3.0-42&field3=buildstamp&compare3=61&value3=20200115-1644-Experimental)
Then we'll be able to move the PR testing from GCC 7.2.0 to this GCC 8.3.0 build.

**Comment by etphipp** [comment]
@ZUUL42, can I get your exact configure script?  So far I have been unable to reproduce the Sacado error.

**Comment by ZUUL42** [comment]
@etphipp, this will show my additions.
[GCC 8.3.0 Diff](https://github.com/trilinos/Trilinos/compare/develop...ZUUL42:gcc_8.3)

**Comment by etphipp** [comment]
Sorry for the stupid question, but can you give me more detail on how to run this?  Do I use the atdm environment for this build, or are there other scripts?

**Comment by etphipp** [comment]
The Sacado error should be gone now as the offending example has been removed.

**Comment by ZUUL42** [comment]
@etphipp, thanks. Looks good.
Down to just [the last error](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=61&value1=ascic166&field2=buildname&compare2=61&value2=PR-gcc_8.3-test-Trilinos_pullrequest_gcc_8.3.0-43&field3=buildstamp&compare3=61&value3=20200127-1848-Experimental) coming from @trilinos/seacas (#6605).

**Comment by ZUUL42** [comment]
I'm going to push forward with this, shifting to package by package WError flags, leaving @trilinos/seacas and @trilinos/muelu without WError protection ... for now.
Here is the latest [CDash](https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=3&showfilters=1&filtercombine=and&field1=site&compare1=61&value1=ascic166&field2=buildname&compare2=61&value2=PR-gcc_8.3-test-Trilinos_pullrequest_gcc_8.3.0-59&field3=buildstamp&compare3=61&value3=20200212-2054-Experimental) that displays those issues.

**Comment by lucbv** [comment]
@ZUUL42 that's a bit disappointing, I have spent time to help with this and of course things just keep slipping in as developers do not know well when they break the requirement of you new build...
To avoid wasting significant amount of developer time, please use this approached as outlined by @bartlettroscoe a while back.
I'll again take the time to fix the issue in MueLu and let you know when my changes are merged and you can turn on the flag everywhere but for seacas

**Comment by bartlettroscoe** [comment]
> To avoid wasting significant amount of developer time, please use this approached as outlined by @bartlettroscoe a while back.

To summarize what I suggested [above](https://github.com/trilinos/Trilinos/issues/6295#issuecomment-568211573) ...

Why not just merge the files:

```
cmake/std/PullRequestLinuxGCC8.3.0TestingSettings.cmake
cmake/std/sems/PullRequestGCC8.3.0TestingEnv.sh
```

from the branch:

* https://github.com/trilinos/Trilinos/compare/develop...ZUUL42:gcc_8.3

to the 'develop' branch ASAP?  There is zero harm in having them in the 'develop' branch and that would make it trivial for developers to reproduce this build on the tip of 'develop'.

They you should set up a nightly build for this configuration with -Werror set.  That way, it works against the tip of 'develop' and people can watch every day to see the errors go away.

Really you should add this as a PR build and set -Werror on packages that are already clean to make sure they don't backslide as well.

I think that is how one wants to do this sort of thing.  Otherwise, this build will never get clean.

**Comment by ZUUL42** [comment]
@lucbv, ya, I wished I had moved this to the PR build a bit sooner. I was hoping to get it set globally upon introduction, but didn't quite get there. I can at least get this into the PR testing set _now_ so that no _more_ packages that are clean in 8.3.0 can backslide further. Similar to how I did 7.2.0 WError, but we're starting closer to the finish line.
@bartlettroscoe, I was aiming to add it to the PR build set on Monday with WError set by package. I was waiting for some last tests to finish before I sent out an email notifying everyone.
I hadn't planned to put a nightly in place that would report on those 2 issues, but I'll look into setting it up.
